### PR TITLE
Store event managers in new db

### PR DESCRIPTION
**Pull request by ThiefMaster**
This is still work-in-progress.

### CODE REVIEW

File: indico/MaKaC/common/indexes.py [hunk 1]
```diff
@@ -1226,9 +1211,7 @@ class IndexesHolder(ObjectHolder):
     counterName = None
     __allowedIdxs = ["calendar", "category", "categoryDate",
                      "categoryDateAll", "categoryName", "conferenceTitle",
-                     "pendingSubmitters", "pendingConfSubmitters",
-                     "pendingConfSubmittersTasks", "pendingConfManagers",
-                     "pendingConfManagersTasks", "pendingSubmittersTasks", "pendingManagers",
+                     "pendingSubmitters", "pendingSubmittersTasks", "pendingManagers",
```
Comment (pferreir) [inline indico/MaKaC/common/indexes.py]: I have the feeling `pendingSubmittersTasks` is not used.

File: indico/MaKaC/conference.py [hunk 2]
```diff
@@ -1882,17 +1882,13 @@ class Conference(CommonObjectBase, Locatable):
...
         """Class constructor. Initialise the class attributes to the default
             values.
            Params:
             confData -- (Dict) Contains the data the conference object has to
                 be initialised to.
         """
-        #IndexedObject.__init__(self)
-        if creator == None:
-            raise MaKaCError( _("A creator must be specified when creating a new Event"), _("Event"))
-        self.__creator = creator
```
Comment (pferreir) [inline indico/MaKaC/conference.py]: Why not?

Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: https://github.com/indico/indico/pull/2055/files#diff-a56367552dee2cb208d8b8f8cadd8dc2R4243

File: indico/MaKaC/conference.py [hunk 3]
```diff
@@ -3582,18 +3572,14 @@ def canAccess( self, aw ):
         if has_request_context() and self.__ac.isHarvesterIP(request.remote_addr):
             return True
 
-        # Managers have always access
-        if self.canModify(aw):
-            return True
-
```
Comment (pferreir) [inline indico/MaKaC/conference.py]: Why?

Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: Moved it down to avoid calling `canModify` in case the event is public and has no IP restriction (no need to trigger SQL queries from canModify/can_manage in that case)

File: indico/MaKaC/conference.py (hunk 2 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: Ah indeed.  But then why keep the `creator` kwarg?

Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: Yeah.. I guess I can get rid of it. the constructor isn't called in that many places anyway

File: indico/MaKaC/conference.py (hunk 3 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: :+1:

File: indico/core/db/sqlalchemy/protection.py [hunk 4]
```diff
@@ -125,12 +129,212 @@ def can_access(self, user, acl_attr='acl', legacy_method='canAccess', allow_admi
...
+
+        :param user: The :class:`.User` to check. May be None if the
+                     user is not logged in.
+        :param: role: The management role that is needed for the
+                      check to succeed.  If not specified, full
+                      management privs are required.  May be set to
+                      the string ``'ANY'`` to check if the user has
+                      any management privileges.  Full management
+                      privs always grant access even if the role is
+                      not granted explicitly.
```
Comment (pferreir) [inline indico/core/db/sqlalchemy/protection.py]: > Full management privs always grant access even if the role is not granted explicitly.
What do you mean with this?

File: indico/MaKaC/conference.py (hunk 2 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: :cool:

File: indico/core/db/sqlalchemy/protection.py (hunk 4 above)
Comment (ThiefMaster) [inline indico/core/db/sqlalchemy/protection.py]: `event.can_manage(user, 'registration')` will succeed if the guy has `full_access=True` on his ACL entry, even if he doesn't have the 'registration' role. Basically `full_access` is like superuser, unless `explicit` is specified.

File: indico/MaKaC/conference.py [hunk 5]
```diff
@@ -3607,91 +3593,35 @@ def canKeyModify(self):
...
-            if owner.canUserModify( av ):
-                return True
-        return False
+    @unify_user_args
+    def canUserModify(self, user, role=None):
+        return self.as_event.can_manage(user, role=role)
 
-    @unify_user_args(legacy=True)
-    def canModify(self, aw_or_user):
+    def canModify(self, aw_or_user, role=None):
```
Comment (pferreir) [inline indico/MaKaC/conference.py]: Why add the `role` kwarg here if this is a legacy API? As long as `can_manage` support roles, why not force full privileges here?

File: indico/MaKaC/conference.py [hunk 6]
```diff
@@ -3607,91 +3593,35 @@ def canKeyModify(self):
...
-
-    def isRegistrar(self, av):
-        if av is None:
-            return False
-        try:
-            return any(principal.containsUser(av) for principal in self.getRegistrarList())
-        except AttributeError:
-            return False
+    def getManagerList(self):
+        managers = sorted([x.principal for x in self.as_event.acl_entries if x.has_management_role()],
```
Comment (pferreir) [inline indico/MaKaC/conference.py]: > `has_management_role()`

I know this can take an arg, but it looks so weird like this.

File: indico/MaKaC/conference.py (hunk 5 above)
Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: https://github.com/indico/indico/pull/2055/files#diff-ab17c90821c50cfe042ab9e5e927e72eR827

That way you can just set e.g. `ROLE = 'registration'` in a RH inheriting from the old `RHConferenceModifBase` RH without having tons of extra logic there to use a different method depending on the object type.

File: indico/MaKaC/conference.py (hunk 6 above)
Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: What about going for an explicit `has_management_role(full_access=True)`? Might be clearer anyway

File: indico/MaKaC/conference.py (hunk 5 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: Wouldn't it be possible to use `can_manage` there? We know we will always have an access wrapper over there... or am I overlooking some of the logic?

File: indico/MaKaC/conference.py (hunk 6 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: I don't like that you'd have to specify a different argument. I think the first version is a lesser evil :confused:

File: indico/core/db/sqlalchemy/protection.py (hunk 4 above)
Comment (pferreir) [inline indico/core/db/sqlalchemy/protection.py]: Ah, OK, I thought that was already implicit. I would rephrase it to something like `Full management privileges are equivalent to specifying all possible roles.`

File: indico/MaKaC/conference.py (hunk 5 above)
Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: It'd need an `isinstance(self._target, Conference)` or `hasattr(self._target, 'can_manage')` check instead of just passing the role if it's defined assuming the object handles it.

But checking for `can_manage` is probably cleaner indeed.

File: indico/MaKaC/conference.py [hunk 7]
```diff
@@ -3798,11 +3728,14 @@ def clone( self, startDate, options, eventManager=None, userPerformingClone = No
...
             conf.setModifKey(self.getModifKey())
         # Access Control cloning
-        if options.get("access",False) :
+        if options.get("access", False):
             conf.setProtection(self.getAccessController()._getAccessProtection())
-            for mgr in self.getManagerList() :
-                conf.grantModification(mgr)
-            for user in self.getAllowedToAccessList() :
+            from indico.modules.events.models.principals import EventPrincipal
+            cols = {column.key for column in inspect(EventPrincipal).column_attrs} - {'event_id', 'id'}
```
Comment (pferreir) [inline indico/MaKaC/conference.py]: There's really no better way to do this?

File: indico/MaKaC/conference.py (hunk 5 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: It looks to me like a better option than "upgrading" a legacy API.

File: indico/MaKaC/conference.py (hunk 7 above)
Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: I could probably add a util function such as `get_columns(EventPrincipal, exclude={'id', 'event_id'})` since we'll probably do this in more places when cloning things.

File: indico/MaKaC/participant.py [hunk 8]
```diff
@@ -448,16 +448,8 @@ def addPendingParticipant(self, participant):
...
-            creator=self._conference.getCreator()
-            if isinstance(creator, AvatarUserWrapper):
-                toList.append(creator.getEmail())
-            for manager in self._conference.getAccessController().getModifierList() :
-                if isinstance(manager, AvatarUserWrapper):
-                    toList.append(manager.getEmail())
-
             data = {}
-            data["toList"] = toList
+            data["toList"] = self._conference.all_manager_emails
```
Comment (pferreir) [inline indico/MaKaC/participant.py]: Unrelated, but this is awful. We shouldn't use "To:" here. At most a "BCC".

Comment (ThiefMaster) [inline indico/MaKaC/participant.py]: When emailing all managers I think having them all in To is not that bad - it's not likely to be a completely unrelated bunch of people and being able to reply-all on such emails sounds somewhat useful

File: indico/MaKaC/webinterface/pages/conferences.py [hunk 9]
```diff
@@ -661,22 +662,22 @@ def _extractInfoForButton(self, item):
...
-            if item.getAccessController().canUserSubmit(self._rh._aw.getUser()):
+            elif item.as_event.can_manage(session.user, 'submit'):
                 info["minutesLink"] = True
                 info["materialLink"] = True
 
         elif itemType == 'Session':
-            session = item.getSession()
-            info['parentProtection'] = session.getAccessController().isProtected()
-            if session.canModify(self._rh._aw) or session.canCoordinate(self._rh._aw):
+            sess = item.getSession()
```
Comment (pferreir) [inline indico/MaKaC/webinterface/pages/conferences.py]: Whoops. Had never realized `flask.session` causes shadowing issues all over.

Comment (ThiefMaster) [inline indico/MaKaC/webinterface/pages/conferences.py]: Luckily not that many places - I think during the initial flask migration there were only 2 or 3 conflicts in the whole codebase

File: indico/MaKaC/conference.py (hunk 7 above)
Comment (pferreir) [inline indico/MaKaC/conference.py]: It's prone to mistakes if someone e.g. adds a new column. Could we make it more explicit somehow (whitelisting more than blacklisting)?

File: indico/MaKaC/participant.py (hunk 8 above)
Comment (pferreir) [inline indico/MaKaC/participant.py]: Yeah, as you see I've deleted the comment almost immediately, as I realised that.

**Comment by ThiefMaster** [comment]
explicit -> explicit_role or require_role

File: indico/MaKaC/conference.py (hunk 7 above)
Comment (ThiefMaster) [inline indico/MaKaC/conference.py]: I think with whitelisting we are more likely to forget to clone something... besides fkey/pkey columns it usually makes sense to clone everything.
